test/e2e/framework/networking_utils.go

```diff
@@ -708,197 +707,261 @@ func CheckReachabilityFromPod(expectToBeReachable bool, timeout time.Duration, n
+	hostPort := net.JoinHostPort(host, strconv.Itoa(port))
+	url := fmt.Sprintf("udp://%s", hostPort)
+
+	ret := UDPPokeResult{}
+
+	// Sanity check inputs, because it has happened.  These are the only things
+	// that should hard fail the test - they are basically ASSERT()s.
+	if host == "" {
+		Failf("Got empty host for UDP poke (%s)", url)
+		return ret
 	}
-
-	if !strings.Contains(string(buf), expect) {
-		return false, fmt.Errorf("Failed to retrieve %q, got %q", expect, string(buf))
+	if port == 0 {
+		Failf("Got port==0 for UDP poke (%s)", url)
+		return ret
 	}
 
-	Logf("Successfully reached %v", uri)
-	return true, nil
-}
-
-func TestNotReachableUDP(ip string, port int, request string) (bool, error) {
-	ipPort := net.JoinHostPort(ip, strconv.Itoa(port))
-	uri := fmt.Sprintf("udp://%s", ipPort)
-	if ip == "" {
-		Failf("Got empty IP for reachability check (%s)", uri)
-		return false, nil
-	}
-	if port == 0 {
-		Failf("Got port==0 for reachability check (%s)", uri)
-		return false, nil
+	// Set default params.
+	if params == nil {
+		params = &UDPPokeParams{}
 	}
 
-	Logf("Testing UDP non-reachability of %v", uri)
+	Logf("Poking %v", url)
 
-	con, err := net.Dial("udp", ipPort)
+	con, err := net.Dial("udp", hostPort)
 	if err != nil {
-		Logf("Confirmed that %s is not reachable", uri)
-		return true, nil
+		ret.Status = UDPError
+		ret.Error = err
+		Logf("Poke(%q): %v", url, err)
+		return ret
 	}
 
 	_, err = con.Write([]byte(fmt.Sprintf("%s\n", request)))
 	if err != nil {
-		Logf("Confirmed that %s is not reachable", uri)
-		return true, nil
+		ret.Error = err
+		neterr, ok := err.(net.Error)
+		if ok && neterr.Timeout() {
+			ret.Status = UDPTimeout
+		} else if strings.Contains(err.Error(), "connection refused") {
+			ret.Status = UDPRefused
+		} else {
+			ret.Status = UDPError
+		}
+		Logf("Poke(%q): %v", url, err)
+		return ret
 	}
 
-	var buf []byte = make([]byte, 1)
-
-	err = con.SetDeadline(time.Now().Add(3 * time.Second))
-	if err != nil {
-		return false, fmt.Errorf("Failed to set deadline: %v", err)
+	if params.Timeout != 0 {
+		err = con.SetDeadline(time.Now().Add(params.Timeout))
+		if err != nil {
+			ret.Status = UDPError
+			ret.Error = err
+			Logf("Poke(%q): %v", url, err)
+			return ret
+		}
 	}
 
+	bufsize := len(params.ResponseContains)
+	if bufsize == 0 {
+		bufsize = 4096
+	}
+	var buf []byte = make([]byte, bufsize)
 	_, err = con.Read(buf)
 	if err != nil {
-		Logf("Confirmed that %s is not reachable", uri)
-		return true, nil
+		ret.Error = err
+		neterr, ok := err.(net.Error)
+		if ok && neterr.Timeout() {
+			ret.Status = UDPTimeout
+		} else if strings.Contains(err.Error(), "connection refused") {
+			ret.Status = UDPRefused
+		} else {
+			ret.Status = UDPError
+		}
+		Logf("Poke(%q): %v", url, err)
+		return ret
 	}
+	ret.Response = buf
 
-	return false, nil
+	if params.ResponseContains != "" && !strings.Contains(string(buf), params.ResponseContains) {
```

Comment:
It only read `len(params.ResponseContains)` bytes into `buf`, so this isn't really `ResponseContains`, it's `ResponsePrefix`.

But maybe (matching the assumptions on the writing side) you should rename it to just `Response`, and then pass a buffer of length `len(params.Response)+2` to `Read()`, and check that the returned length is `len(params.Response)+1` and that it matches `params.Response+"\n"`. (ie, the server sent back exactly `Response` plus a newline, and nothing more.)

Reply:
Why bother with newline, just do exact comparison?  Will fix

Reply:
Oh, oops... the code appends a newline to the request string, and the original `TestReachableUDP` read back one byte more than the length of the response string but then didn't look at the last byte, so I was assuming that there was a newline in the response too. But I guess there isn't. (And, in fact, the server doesn't care if there's a newline in the request either; it just ignores any trailing whitespace.)